src/app/features/psbt-signer/hooks/match-sats-with-outputs.ts

```diff
@@ -0,0 +1,84 @@
+import { isDefined } from '@shared/utils';
+
+import { PsbtInput } from './use-parsed-inputs';
+import { PsbtOutput } from './use-parsed-outputs';
+
+interface MatchedSatsWithOutput {
+  inputs: PsbtInput[];
+  output: PsbtOutput;
+}
+
+function consolidateSatsWithSameAddress(inputs: PsbtInput[]): PsbtInput[] {
+  const utxo: Record<string, PsbtInput> = {};
+
+  inputs.forEach(({ address, value, ...rest }) => {
+    if (isDefined(utxo[address])) {
+      utxo[address] = {
+        ...utxo[address],
+        value: utxo[address].value + value,
+      };
+    } else {
+      utxo[address] = { address, value, ...rest };
+    }
+  });
+  return Object.values(utxo);
+}
+
+export function matchSatsWithOuputs(
+  inputs: PsbtInput[],
+  outputs: PsbtOutput[]
+): {
+  fee: number;
+  matchedSats: MatchedSatsWithOutput[];
+} {
+  const matchedSatsWithOutput: MatchedSatsWithOutput[] = [];
+  const availableUtxos = [...inputs];
+  // Keep track of the sats available across multiple outputs
+  let remainingSats = 0;
+
+  for (const output of outputs) {
+    // Note here that the matched sats uses an array, so multiple
+    // utxos can be used to fulfill an output
+    const matched: MatchedSatsWithOutput = {
+      inputs: [],
+      output: output,
+    };
+
+    // Calculate the remaining sats needed for this output
+    let remainingOutputValue = Number(output.value);
+
+    // Iterate through the available sats and add them until the output is fulfilled
+    for (let i = 0; i < availableUtxos.length; i++) {
+      const utxo = availableUtxos[i];
+
+      if (remainingSats === 0) remainingSats = utxo.value;
+
+      if (remainingSats <= remainingOutputValue) {
+        // Use all the sats and subtract from the output value
+        matched.inputs.push(utxo);
+        remainingOutputValue -= remainingSats;
+        remainingSats = 0;
+        availableUtxos.splice(i, 1);
+        i--;
+      } else {
+        // Use the sats, but keep track of the remainder for future outputs
+        remainingSats -= remainingOutputValue;
+        matched.inputs.push(utxo);
+        remainingOutputValue = 0;
+      }
+
+      if (remainingOutputValue === 0) break;
+    }
+
+    matchedSatsWithOutput.push(matched);
+  }
+
+  const matchedSats: MatchedSatsWithOutput[] = matchedSatsWithOutput.map(matched => {
+    return {
+      inputs: consolidateSatsWithSameAddress(matched.inputs),
+      output: matched.output,
+    };
+  });
+
+  return { fee: remainingSats, matchedSats };
```

Comment:
Sats are the value on the input (utxo) ...makes sense to me. Ultimately, we need to know the inputs/outputs that match up bc the input (utxo) funds the output and the tx output is what we query to check for the inscription. I can change the name to use `inputs`? Sats are the input value.

Not sure I understand your suggestion?

Reply:
I think maybe I just have the logic wrong in how inputs and outputs work? I was thinking a utxo is used across multiple outputs until it is spent, but that isn't correct? It can only be used for one output. Probably why doesn't make sense to you? I think I get your suggestion now.

Reply:
> Sats are the value on the input

Maybe we have a different understanding.

I'd describe as: An input is a collection of sats. Afaik, the only reason we ever need to "match", is to know if an inscription (on a sat) goes to a given output. 

If so, then isn't it better to think of it as where does this sat (of this input) go, rather than where this input goes. 
This is a simpler model as a sat-to-output is a one to one relationship, whereas an input to output is a one to many relationship.

Reply:
I think what you describe is what I'm doing, I am using the input value (sats) to do the matching. However, I don't think the input/output is a one-to-many relationship? That is what I'm questioning now about my implementation. I was just doing more reading on it, and I think one input (utxo) is matched to one output. Are you confident input/output is one-to-many bc chatGPT tells me differently... the remainder of each utxo is change.

<img width="843" alt="Screen Shot 2023-07-19 at 12 06 05 PM" src="https://github.com/hirosystems/wallet/assets/6493321/41e633db-91d5-4245-91db-84242e2b507f">

Reply:
And, yes, we are currently using just for inscriptions and the fee (but we can determine the fee with a simple subtraction of outputs from inputs if needed).

Reply:
Redoing bc I over-complicated the matching, brb!

Reply:
Actually, I don't think I did. I think chapGPT is incorrect? It depends really on the sighash type?

Reply:
> I'd describe as: An input is a collection of sats. Afaik, the only reason we ever need to "match", is to know if an inscription (on a sat) goes to a given output.
> 
We don't make a query per sat for the inscription. I am matching up the sat value of the input with the outputs to determine if that utxo is being spent for that output. Then, I'm checking if the utxo itself has an inscription, not per sat.
>
> If so, then isn't it better to think of it as where does this sat (of this input) go, rather than where this input goes. This is a simpler model as a sat-to-output is a one to one relationship, whereas an input to output is a one to many relationship.
> 
I think this is what I'm doing tbh? Except that, if the sighash type isn't ALL, there would be a problem.
> 
> Am I correct in thinking we only "match" to know where an inscription goes? or are there other use cases?
>
Yes, and currently for the fee.

Reply:
> 
> const output = satMatcher.findOutput({ txid, offset });
> ```
> 
I don't understand this, how are you finding a tx output on a PSBT with a `txid` and an `offset`? The output on a PSBT is a `script` (which contains the receiving address) and an `amount`. 🤔

Reply:
Also, I followed a similar pattern I saw others doing, where I parse the PSBT inputs and put some extra data on it to use in the UI, type `PsbtInput`, this already has the inscription data, if there is one. So, when they are matched to outputs, it is easy to know if there is an inscription on the transferring `input` (utxo) or receiving `output`.

Reply:
I did some renaming today, and added some more tests, to help clarify what is happening here. Although it isn't checking inscriptions per sat, I do feel like it is working at the utxo level to detect the inscription and which output is receiving it. I think this makes the assumption the tx is being constructed correctly though.

Reply:
Yep, I get everything you are saying and agree. I don't think what you are suggesting is far off what I have done. The totals are being done separately. The only prob is the issue with the example you show. I think there is just a gap in my understanding when you say `finds corresponding output that has it` ...I have no idea what you mean there if you don't do matching?

Reply:
Another example to better highlight what we discussed on call

<img width="910" alt="image" src="https://github.com/hirosystems/wallet/assets/1618764/bc0222d7-6235-40b7-9ebc-456f289c7f76">